pkg/gen/ami/ami.go

```diff
@@ -0,0 +1,206 @@
+	channel        string
+	chinaDomain    string
+	dir            string
+	minimumVersion string
+	primaryDomain  string
+}
+
+func NewAMI(config Config) (*AMI, error) {
+	err := config.Validate()
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	f := &AMI{
+		arch:           config.Arch,
+		channel:        config.Channel,
+		chinaDomain:    config.ChinaDomain,
+		dir:            config.Dir,
+		minimumVersion: config.MinimumVersion,
+		primaryDomain:  config.PrimaryDomain,
+	}
+
+	return f, nil
+}
+
+type amiInfo struct {
+	Name string `json:"name"`
+	PV   string `json:"pv"`
+	HVM  string `json:"hvm"`
+}
+
+type amiInfoList struct {
+	AMI []amiInfo `json:"amis"`
+}
+
+func (f *AMI) GetInput(ctx context.Context) (input.Input, error) {
+	templateData, err := f.getAMI(ctx)
+	if err != nil {
+		return input.Input{}, microerror.Mask(err)
+	}
+	i := input.Input{
+		Path:         filepath.Join(f.dir, "ami.go"),
+		TemplateBody: amiTemplate,
+		TemplateData: templateData,
+	}
+
+	return i, nil
+}
+
+func (f *AMI) getAMI(ctx context.Context) (map[string]interface{}, error) {
+	var versions []string
+	{
+		url := fmt.Sprintf("https://%s.release.%s/%s/", f.channel, f.primaryDomain, f.arch)
+		fmt.Println("scraping", url)
+		response, err := http.Get(url) //nolint:gosec
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		versions, err = scrapeVersions(response.Body)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	mergedAMI := map[string]map[string]string{}
+	for _, version := range versions {
+		if semver.MustParse(version).LessThan(semver.MustParse(f.minimumVersion)) {
+			continue
+		}
+		url := fmt.Sprintf("https://%s.release.%s/%s/%s/flatcar_production_ami_all.json", f.channel, f.primaryDomain, f.arch, version)
+		fmt.Println("scraping", url)
+		response, err := http.Get(url) //nolint:gosec
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		if response.StatusCode == 403 {
+			continue // not found, keep going
+		}
+		mergedAMI[version], err = scrapeVersionAMI(response.Body)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	for version := range mergedAMI {
+		url := fmt.Sprintf("https://%s/%s/%s/%s.json", f.chinaDomain, f.channel, f.arch, version)
+		fmt.Println("scraping", url)
+		response, err := http.Get(url) //nolint:gosec
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		if response.StatusCode == 403 {
+			continue
+		}
+		chinaVersionAMI, err := scrapeVersionAMI(response.Body)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		for region, image := range chinaVersionAMI {
+			mergedAMI[version][region] = image
+		}
+	}
+
+	result, err := json.MarshalIndent(mergedAMI, "", "  ")
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	templateData := map[string]interface{}{
+		"AMIInfoString": fmt.Sprintf("`%s`", result),
+		"Package":       "key",
+	}
+
+	return templateData, nil
+}
+
+func scrapeVersions(source io.Reader) ([]string, error) {
```

Comment:
This is some serious shit.

Reply:
I always knew my Python HTML scraping experience from college would be useful one day.